examples/bsoncxx/builder_stream_fancy.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <map>
+#include <vector>
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/builder/stream/array.hpp>
+#include <bsoncxx/json.hpp>
+
+using namespace bsoncxx;
+
+// concatenates arbitrary ranges into an array context
+template <typename begin_t, typename end_t>
+class range_array_appender {
+   public:
+    range_array_appender(begin_t begin, end_t end) : _begin(begin), _end(end) {
+    }
+
+    void operator()(bsoncxx::builder::stream::array_context<> ac) const {
+        for (auto iter = _begin; iter != _end; ++iter) {
```

Comment:
See the new C++17 for loop body:

```
{
    auto && __range = range_expression ; 
    auto __begin = begin_expr ;
    auto __end = end_expr ;
    for ( ; __begin != __end; ++__begin) { 
        range_declaration = *__begin; 
        loop_statement 
    } 
} 
```

I think the point is that end sentinels shouldn't have to be as fat as active iterators just to make the type system happy